### 1. Example 3.5: Compare coefficients from lm() with $\hat{\mu_i}$ & $\hat{\alpha_i}$

輸入資料以及提取變數

In [5]:
resin = read.table("http://www.stat.umn.edu/~gary/book/fcdae.data/exmpl3.2", header = TRUE)
y = resin$y
temp = resin$temp

建立模型

In [6]:
g = lm(y ~ factor(temp), data = resin)
summary(g)


Call:
lm(formula = y ~ factor(temp), data = resin)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.22667 -0.03667  0.00250  0.03125  0.20333 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)    1.93250    0.03387  57.055  < 2e-16 ***
factor(temp)2 -0.30375    0.04790  -6.341 4.06e-07 ***
factor(temp)3 -0.55500    0.04790 -11.586 5.49e-13 ***
factor(temp)4 -0.73821    0.04958 -14.889 6.13e-16 ***
factor(temp)5 -0.87583    0.05174 -16.928  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.0958 on 32 degrees of freedom
Multiple R-squared:  0.9233,	Adjusted R-squared:  0.9138 
F-statistic: 96.36 on 4 and 32 DF,  p-value: < 2.2e-16


從 inClass2 的題目中可以算出

In [7]:
anova(g)
mu_hat = mean(y)
group_mean = tapply(y, temp, mean)
alpha_hat = group_mean - mu_hat
table = cbind(group_mean, alpha_hat)
table

Df Sum Sq    Mean Sq     F value  Pr(>F)      
factor(temp)  4 3.5376321 0.884408016 96.36296 2.241949e-17
Residuals    32 0.2936923 0.009177883       NA           NA

group_mean alpha_hat  
1 1.932500    0.46736486
2 1.628750    0.16361486
3 1.377500   -0.08763514
4 1.194286   -0.27084942
5 1.056667   -0.40846847

|組別|$\hat{\mu_i}$|$\hat{\alpha_i}$|
|:---:|:---:|:---:|
|1|1.932500|0.46736486|
|2|1.628750|0.16361486|
|3|1.377500|-0.08763514|
|4|1.194286|-0.27084942|
|5|1.056667|-0.40846847|

其中用 lm() 建立的 linear model 中的係數是以 factor(temp)1 當作<font color=#FF0000>reference</font>，其他 factor 與 reference 的差異，故如下表：

|組別||$\hat{\mu_i}$|
|:---:|:---:|:---:|
|1|1.932500 =|1.932500|
|2|1.932500 - 0.30375 = |1.628750|
|3|1.932500 - 0.55500 = |1.377500|
|4|1.932500 - 0.73821 = |1.194286|
|5|1.932500 - 0.87583 = |1.056667|

再計算 $\bar{y_{..}}$ 

In [8]:
mean(resin[, 2])

[1] 1.465135

剛剛已經從 linear model 內計算出 $\hat{\mu_i}$ 的大小
接下來用$\hat{\mu_i} - \bar{y_{..}}$ 即可得到各組的 $\hat{\alpha_i}$，如下表：

|組別||\hat{\alpha_i}|
|:--:|:--:|:--:|
|1|1.932500-1.465135=|0.46736486|
|2|1.628750-1.465135=|0.16361486|
|3|1.377500-1.465135=|-0.08763514|
|4|1.194286-1.465135=|-0.27084942|
|5|1.056667-1.465135=|-0.40846847|

### 2. Do Exercise 3.5: obtain also treatment effects

讀入資料

In [9]:
angle = read.table("http://www.stat.umn.edu/~gary/book/fcdae.data/ex3.5", header = T)
angle

trt deg
1  1   140
2  1   138
3  1   140
4  1   138
5  1   142
6  2   140
7  2   150
8  2   120
9  2   128
10 2   130
11 3   118
12 3   130
13 3   128
14 3   118
15 3   118

建立模型

In [10]:
model = lm(deg ~ factor(trt), data = angle)
summary(model)


Call:
lm(formula = deg ~ factor(trt), data = angle)

Residuals:
   Min     1Q Median     3Q    Max 
 -13.6   -4.4   -1.6    4.0   16.4 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept)   139.600      3.410  40.941 2.92e-14 ***
factor(trt)2   -6.000      4.822  -1.244  0.23716    
factor(trt)3  -17.200      4.822  -3.567  0.00387 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 7.625 on 12 degrees of freedom
Multiple R-squared:  0.5221,	Adjusted R-squared:  0.4425 
F-statistic: 6.555 on 2 and 12 DF,  p-value: 0.01191


寫出 anova table

In [11]:
anova(model)

Df Sum Sq   Mean Sq   F value  Pr(>F)    
factor(trt)  2 762.1333 381.06667 6.555046 0.01191236
Residuals   12 697.6000  58.13333       NA         NA

先算group_mean再減去整組資料的總平均，也就是 deg 的總平均，用以計算 treatment effects

In [12]:
group_mean = tapply(angle$deg, angle$trt, mean)
group_mean

1     2     3 
139.6 133.6 122.4

In [13]:
deg_mean = mean(angle$deg)
treatment_effects = group_mean - deg_mean
treatment_effects

1         2         3 
 7.733333  1.733333 -9.466667

即可得到各組的 treatment effects